File: Source/TagGrouper.cs
namespace NhlStenden.TagBag;

/// <summary>
/// Represents a class that groups specified tags in files.
/// </summary>
public class TagGrouper : TagBase
{
    /// <summary>
    /// Executes the tag grouping process on the specified files.
    /// </summary>
    /// <param name="args">A list of arguments where the first argument is the directory path and subsequent arguments are the groups of tags.</param>
    public static void Execute(IList<string> args)
    {
        var fileNames = GetTagFiles(args[0]);
        var groups = args.Count >= 2 ? [.. args.ToArray()[1..]] : new List<string>();

        foreach (var fileName in fileNames)
            ProcessFile(fileName, groups);
    }

    /// <summary>
    /// Processes the specified file to group specified tags.
    /// </summary>
    /// <param name="fileName">The name of the file to process.</param>
    /// <param name="groups">A list of groups of tags to be grouped in the file.</param>
    static void ProcessFile(string fileName, List<string> groups)
    {
        var lines = File.ReadAllLines(fileName).ToList();

        for (var i = 0; i < lines.Count; i++)
        {
            var line = lines[i];
            var newLine = new List<string>();

            var items = TagHelper.SplitIntoTags(line).ToList();
            foreach (var group in groups)
            {
                var groupItems = TagHelper.SplitIntoTags(group);
                var groupIndices = IndexOf(items, groupItems).OrderByDescending(x => x).ToList();
                if (ValidIndices(groupIndices))
                {
                    // regroup items into new version of items
                    foreach (var groupIndex in groupIndices)
                    {
                        items.RemoveAt(groupIndex);
                    }
                    items.InsertRange(groupIndices[^1], groupItems);
                }
            }

            newLine = items.ToList();
            line = string.Join(", ", newLine);
            lines[i] = line;
        }

        File.WriteAllLines(fileName, lines);
    }

    /// <summary>
    /// Finds the indices of the specified needles in the haystack.
    /// </summary>
    /// <typeparam name="T">The type of elements in the lists.</typeparam>
    /// <param name="haystack">The list to search in.</param>
    /// <param name="needles">The list of elements to find.</param>
    /// <returns>A list of indices where the needles are found in the haystack.</returns>
    static IList<int> IndexOf<T>(IList<T> haystack, IList<T> needles)
    {
        var result = new List<int>();
        foreach (var needle in needles)
        {
            result.Add(haystack.IndexOf(needle));
        }
        return result;
    }

    /// <summary>
    /// Validates the indices to ensure they are all non-negative.
    /// </summary>
    /// <param name="indices">The list of indices to validate.</param>
    /// <returns>True if all indices are non-negative; otherwise, false.</returns>
    static bool ValidIndices(IList<int> indices)
    {
        foreach (var index in indices)
            if (index < 0) return false;

        return true;
    }
}

File: tagbag/tag_grouper.py
from .tag_base import TagBase
from .tag_helper import TagHelper


from pathlib import Path
from typing import List, Sequence


class TagGrouper(TagBase):
    """
    Regroup specified tag groups inside lines.

    Example:
      groups: ["a, b, c", "x, y"]
      line:   "c, a, b, z, y, x"
      result: "... a, b, c ..." and "... x, y ..." when all group elements are present
    """

    @staticmethod
    def execute(args: Sequence[str]) -> None:
        """
        args[0]: directory path
        args[1:]: group strings of comma separated tags
        """
        dir_path = args[0]
        groups = list(args[1:]) if len(args) >= 2 else []

        for name in TagBase.get_tag_files(dir_path):
            TagGrouper._process_file(Path(name), groups)

    @staticmethod
    def _process_file(file_path: Path, groups: List[str]) -> None:
        lines = file_path.read_text(encoding="utf-8").splitlines()
        out_lines: List[str] = []

        for line in lines:
            items = [t.strip() for t in TagHelper.split_into_tags(line)]
            cur = list(items)

            for group in groups:
                group_items = [t.strip() for t in TagHelper.split_into_tags(group)]
                indices = TagGrouper._indices_of(cur, group_items)
                indices_sorted_desc = sorted(indices, reverse=True)
                if TagGrouper._valid_indices(indices_sorted_desc):
                    for idx in indices_sorted_desc:
                        cur.pop(idx)
                    insert_at = min(indices_sorted_desc)
                    cur[insert_at:insert_at] = group_items

            out_lines.append(", ".join(cur))

        file_path.write_text("\n".join(out_lines), encoding="utf-8")

    @staticmethod
    def _indices_of(haystack: List[str], needles: List[str]) -> List[int]:
        """
        Return the indices of each needle (first occurrence) in haystack.
        Missing needles give -1 which invalidates the group.
        """
        return [haystack.index(n) if n in haystack else -1 for n in needles]

    @staticmethod
    def _valid_indices(indices: List[int]) -> bool:
        return all(i >= 0 for i in indices)